server/src/main/java/org/elasticsearch/action/search/SearchTransportService.java

```diff
@@ -96,7 +96,8 @@ public SearchTransportService(Settings settings, TransportService transportServi
     }
 
     public Map<String, Long> getClientConnections() {
-        return Collections.unmodifiableMap(clientConnections);
+        // create a (weakly consistent) snapshot of the map so that it does not get modified when serializing it for stats reporting
+        return Collections.unmodifiableMap(new HashMap<>(clientConnections));
```

Comment:
I think this is subject to the same race, concurrent modification during the copy? It won’t manifest as an inconsistent snapshot of size and number of entries when iterating for serialization, but instead a concurrent modification exception thrown during the copy?

Reply:
See https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html

> Similarly, Iterators, Spliterators and Enumerations return elements reflecting the state of the hash table at some point at or since the creation of the iterator/enumeration. They do not throw ConcurrentModificationException

Reply:
The problem is I did not realize the backing map was a concurrent hash map (the perils of reviewing on mobile).